full-stack-intro/client/src/components/Todos.tsx

```diff
@@ -0,0 +1,155 @@
+/* eslint-disable @typescript-eslint/no-unused-vars -- Remove me */
+/* eslint-disable @typescript-eslint/no-empty-function -- Remove me */
+import { useEffect, useState } from 'react';
+import { PageTitle } from './PageTitle';
+import { TodoList } from './TodoList';
+import { TodoForm } from './TodoForm';
+
+export type UnsavedTodo = {
+  task: string;
+  isCompleted: boolean;
+};
+export type Todo = UnsavedTodo & {
+  todoId: number;
+};
+
+export function Todos() {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<unknown>();
+
+  /* Implement useEffect to fetch all todos. Hints are at the bottom of the file. */
+  useEffect(() => {
+    const fetchTodos = async () => {
+      try {
+        const response = await fetch('/api/todos');
+        if (!response.ok)
+          throw new Error(`Response status: ${response.status}`);
+        const todosJSON = (await response.json()) as Todo[];
+        console.log('todosJSON', todosJSON);
+        setTodos(todosJSON);
+      } catch (error) {
+        setError(error);
+      } finally {
+        setIsLoading(false);
+      }
+    };
+    fetchTodos();
+  }, []);
+
+  /* Implement addTodo to add a new todo. Hints are at the bottom of the file. */
+  async function addTodo(newTodo: UnsavedTodo) {
+    console.log('newTodo', newTodo);
+    try {
+      const response = await fetch('/api/todos', {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(newTodo),
+      });
+      if (!response.ok) throw new Error(`Response status: ${response.status}`);
+      const todo = (await response.json()) as Todo;
+      setTodos([...todos, todo]);
+    } catch (error) {
+      setError(error);
+    }
+  }
+
+  /* Implement toggleCompleted to toggle the completed state of a todo. Hints are at the bottom of the file. */
+  async function toggleCompleted(todo: Todo) {
+    try {
+      const response = await fetch('/api/todos/' + todo.todoId, {
+        method: 'PUT',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify({ ...todo, isCompleted: !todo.isCompleted }),
+      });
+      const updatedTodo = (await response.json()) as Todo;
+      const todosUpdated = todos.map((todo) =>
+        todo.todoId === updatedTodo.todoId ? (todo = updatedTodo) : todo
+      );
```

Comment:
You should also check if `response.ok` is true or not and throw an error otherwise